pl_examples/domain_templates/computer_vision_fine_tuning.py

```diff
@@ -0,0 +1,440 @@
+        _make_trainable(module=child)
+
+
+def filter_params(module: torch.nn.Module,
+                  train_bn: bool = True) -> Generator:
+    """Yields the trainable parameters of a given module.
+
+    Args:
+        module: A given module
+        train_bn: If True, leave the BatchNorm layers in training mode
+
+    Returns:
+        Generator
+    """
+    children = list(module.children())
+    if not children:
+        if not (isinstance(module, BN_TYPES) and train_bn):
+            for param in module.parameters():
+                if param.requires_grad:
+                    yield param
+    else:
+        for child in children:
+            for param in filter_params(module=child, train_bn=train_bn):
+                yield param
+
+
+def _unfreeze_and_add_param_group(module: torch.nn.Module,
+                                  optimizer: Optimizer,
+                                  lr: Optional[float] = None,
+                                  train_bn: bool = True):
+    """Unfreezes a module and adds its parameters to an optimizer."""
+    _make_trainable(module)
+    params_lr = optimizer.param_groups[0]['lr'] if lr is None else float(lr)
+    optimizer.add_param_group(
+        {'params': filter_params(module=module, train_bn=train_bn),
+         'lr': params_lr / 10.,
+         })
+
+
+#  --- Pytorch-lightning module ---
+
+
+class TransferLearningModel(pl.LightningModule):
+    """Transfer Learning with pre-trained ResNet50.
+
+    Args:
+        hparams: Model hyperparameters
+        dl_path: Path where the data will be downloaded
+    """
+    def __init__(self,
+                 hparams: argparse.Namespace,
+                 dl_path: Union[str, Path]) -> None:
+        super().__init__()
+        self.hparams = hparams
+        self.dl_path = dl_path
+        self.__build_model()
+
+    def __build_model(self):
+        """Define model layers & loss."""
+
+        # 1. Load pre-trained network:
+        model_func = getattr(models, self.hparams.backbone)
+        backbone = model_func(pretrained=True)
+
+        _layers = list(backbone.children())[:-1]
+        self.feature_extractor = torch.nn.Sequential(*_layers)
+        freeze(module=self.feature_extractor, train_bn=self.hparams.train_bn)
+
+        # 2. Classifier:
+        _fc_layers = [torch.nn.Linear(2048, 256),
+                      torch.nn.Linear(256, 32),
+                      torch.nn.Linear(32, 1)]
+        self.fc = torch.nn.Sequential(*_fc_layers)
+
+        # 3. Loss:
+        self.loss_func = F.binary_cross_entropy_with_logits
+
+    def forward(self, x):
+        """Forward pass. Returns logits."""
+
+        # 1. Feature extraction:
+        x = self.feature_extractor(x)
+        x = x.squeeze(-1).squeeze(-1)
+
+        # 2. Classifier (returns logits):
+        x = self.fc(x)
+
+        return x
+
+    def loss(self, labels, logits):
+        return self.loss_func(input=logits, target=labels)
+
+    def train(self, mode=True):
```

Comment:
what is the `mode` for? could it be more descriptive?

Reply:
See https://github.com/pytorch/pytorch/blob/d37a4861b8a5eed3d9a1340484d1efb0f48aa59e/torch/nn/modules/module.py#L1067. This line overrides the `train` method of the Pytorch module. I will add a docstring specifying what `mode` does.

Reply:
you are right, we may rename it...
doe you have suggestion about a better name? @PyTorchLightning/core-contributors

Reply:
I am not sure we can rename it. In the [evaluation loop (L330)](https://github.com/PyTorchLightning/pytorch-lightning/blob/42d5cfc3b056b4c82a77a7cdcb8eafc63a812b67/pytorch_lightning/trainer/evaluation_loop.py#L330), `model.train()` is called and here, `model` refers (if I am not mistaken) to the LightningModule. We want to override this `train` method so that, at the end of the evaluation loop, some parameters (in specific layers) remain frozen (that is, with `requires_grad=False`) if needed.